service/src/main/java/uk/gov/hmcts/reform/fpl/service/ccd/CoreCaseDataService.java

```diff
@@ -15,39 +15,116 @@
 import uk.gov.hmcts.reform.fpl.service.SystemUserService;
 import uk.gov.hmcts.reform.fpl.utils.elasticsearch.MatchQuery;
 
+import java.util.HashMap;
 import java.util.Map;
+import java.util.function.Function;
 
 import static java.util.Collections.emptyMap;
 import static uk.gov.hmcts.reform.fpl.CaseDefinitionConstants.CASE_TYPE;
 import static uk.gov.hmcts.reform.fpl.CaseDefinitionConstants.JURISDICTION;
 
-@Slf4j
 @Service
+@Slf4j
 @RequiredArgsConstructor(onConstructor = @__(@Autowired))
 public class CoreCaseDataService {
     private final AuthTokenGenerator authTokenGenerator;
     private final CoreCaseDataApi coreCaseDataApi;
     private final RequestData requestData;
     private final SystemUserService systemUserService;
+    private final CCDConcurrencyHelper concurrencyHelper;
+
+    public CaseDetails performPostSubmitCallback(Long caseId,
+                                          String eventName,
+                                          Function<CaseDetails, Map<String, Object>> changeFunction) {
+        int retries = 0;
+        while (retries < 3) {
+            try {
+                StartEventResponse startEventResponse = concurrencyHelper.startEvent(caseId, eventName);
+                CaseDetails caseDetails = startEventResponse.getCaseDetails();
+                // Work around immutable maps
+                HashMap<String, Object> caseDetailsMap = new HashMap<>(caseDetails.getData());
+                caseDetails.setData(caseDetailsMap);
+
+                Map<String, Object> updates = changeFunction.apply(startEventResponse.getCaseDetails());
+
+                if (!updates.isEmpty()) {
+                    log.info("Submitting event {} on case {}", eventName, caseId);
+                    concurrencyHelper.submitEvent(startEventResponse, caseId, updates);
+                } else {
+                    log.info("No updates, skipping submit event");
+                }
+                caseDetails.getData().putAll(updates);
+                return caseDetails;
+            } catch (Exception e) {
+                log.error("Failed to create event on ccd", e);
+                retries++;
+            }
+        }
+        log.error("All 3 retries failed");
+        return null;
+    }
 
+    public CaseDetails performPostSubmitCallbackUpdateCase(Long caseId,
+                                                           Function<CaseDetails, Map<String, Object>> changeFunction) {
```

Comment:
Do we need this for internal-change-UPDATE_CASE? doesn't seem to get used that much and in other places we just use performPostSubmitCallback and pass in the event name